### PR TITLE
Sanitize URLs in file fields to handle invalid pipe characters ('|')

### LINKED ISSUE #11876: Illegal character in query: pipe character in file field
### JabRef version

Other (please describe below)

### Operating system

GNU / Linux

### Details on version and operating system

_No response_

### Checked with the latest development build (copy version output from About dialog)

- [x] I made a backup of my libraries before testing the latest development version.
- [x] I have tested the latest development version and the problem persists

### Steps to reproduce the behaviour

I think this exception was trigger by searching full text online, which added a URL to the file field. The important part is that a pipe character in the URL raised an exception that recurred until l I killed JabRef and removed the offending URL from the .bib file.

1. Add a URL containing a pipe (`|`) character to the file field (URLs from GALE ONEFILE, containing the string `&docId=GALE|`)

### Appendix

JabRef 6.0--2024-09-25--91d0709
Linux 6.9.9-1-default amd64 
Java 21.0.2 
JavaFX 23+29

<details>
<summary>Log File</summary>

```
java.lang.IllegalArgumentException: Illegal character in query at index 58: https://go.gale.com/ps/i.do?p=HRCA&u=googlescholar&id=GALE|A653693412&v=2.1&it=r&sid=googleScholar&asid=68b1dec7
	at java.base/java.net.URI.create(Unknown Source)
	at org.jabref@6.0.6325/org.jabref.logic.importer.util.FileFieldParser.convert(Unknown Source)
	at org.jabref@6.0.6325/org.jabref.logic.importer.util.FileFieldParser.parse(Unknown Source)
	at org.jabref@6.0.6325/org.jabref.logic.importer.util.FileFieldParser.parse(Unknown Source)
	at java.base/java.util.Optional.map(Unknown Source)
	at com.tobiasdiez.easybind@2.2.1-SNAPSHOT/com.tobiasdiez.easybind.optional.PreboundOptionalBinding$1.computeValue(Unknown Source)
	at com.tobiasdiez.easybind@2.2.1-SNAPSHOT/com.tobiasdiez.easybind.optional.PreboundOptionalBinding$1.computeValue(Unknown Source)
	at javafx.base@23/javafx.beans.binding.ObjectBinding.get(Unknown Source)
	at javafx.base@23/javafx.beans.binding.ObjectExpression.getValue(Unknown Source)
	at com.tobiasdiez.easybind@2.2.1-SNAPSHOT/com.tobiasdiez.easybind.optional.PreboundOptionalBinding$3.computeValue(Unknown Source)
	at javafx.base@23/javafx.beans.binding.ObjectBinding.get(Unknown Source)
	at javafx.base@23/javafx.beans.binding.ObjectExpression.getValue(Unknown Source)
	at javafx.controls@23/javafx.scene.control.TableCell.updateItem(Unknown Source)
	at javafx.controls@23/javafx.scene.control.TableCell.indexChanged(Unknown Source)
	at javafx.controls@23/javafx.scene.control.IndexedCell.updateIndex(Unknown Source)
	at javafx.controls@23/javafx.scene.control.skin.TableRowSkinBase.updateCells(Unknown Source)
	at javafx.controls@23/javafx.scene.control.skin.TableRowSkinBase.checkState(Unknown Source)
	at javafx.controls@23/javafx.scene.control.skin.TableRowSkinBase.computePrefHeight(Unknown Source)
	at javafx.controls@23/javafx.scene.control.Control.computePrefHeight(Unknown Source)
	at javafx.graphics@23/javafx.scene.Parent.prefHeight(Unknown Source)
	at javafx.graphics@23/javafx.scene.layout.Region.prefHeight(Unknown Source)
	at javafx.controls@23/javafx.scene.control.skin.VirtualFlow.resizeCell(Unknown Source)
	at javafx.controls@23/javafx.scene.control.skin.VirtualFlow.getCell(Unknown Source)
	at javafx.controls@23/javafx.scene.control.skin.VirtualFlow.getOrCreateCellSize(Unknown Source)
	at javafx.controls@23/javafx.scene.control.skin.VirtualFlow.getOrCreateCellSize(Unknown Source)
	at javafx.controls@23/javafx.scene.control.skin.VirtualFlow.recalculateAndImproveEstimatedSize(Unknown Source)
	at javafx.controls@23/javafx.scene.control.skin.VirtualFlow.recalculateAndImproveEstimatedSize(Unknown Source)
	at javafx.controls@23/javafx.scene.control.skin.VirtualFlow.recalculateEstimatedSize(Unknown Source)
	at javafx.controls@23/javafx.scene.control.skin.VirtualFlow.layoutChildren(Unknown Source)
	at javafx.graphics@23/javafx.scene.Parent.layout(Unknown Source)
	at javafx.graphics@23/javafx.scene.Parent.layout(Unknown Source)
	at javafx.graphics@23/javafx.scene.Parent.layout(Unknown Source)
	at javafx.graphics@23/javafx.scene.Parent.layout(Unknown Source)
	at javafx.graphics@23/javafx.scene.Parent.layout(Unknown Source)
	at javafx.graphics@23/javafx.scene.Parent.layout(Unknown Source)
	at javafx.graphics@23/javafx.scene.Parent.layout(Unknown Source)
	at javafx.graphics@23/javafx.scene.Parent.layout(Unknown Source)
	at javafx.graphics@23/javafx.scene.Parent.layout(Unknown Source)
	at javafx.graphics@23/javafx.scene.Parent.layout(Unknown Source)
	at javafx.graphics@23/javafx.scene.Scene.doLayoutPass(Unknown Source)
	at javafx.graphics@23/javafx.scene.Scene$ScenePulseListener.pulse(Unknown Source)
	at javafx.graphics@23/com.sun.javafx.tk.Toolkit.lambda$runPulse$2(Unknown Source)
	at java.base/java.security.AccessController.doPrivileged(Unknown Source)
	at javafx.graphics@23/com.sun.javafx.tk.Toolkit.runPulse(Unknown Source)
	at javafx.graphics@23/com.sun.javafx.tk.Toolkit.firePulse(Unknown Source)
	at javafx.graphics@23/com.sun.javafx.tk.quantum.QuantumToolkit.pulse(Unknown Source)
	at javafx.graphics@23/com.sun.javafx.tk.quantum.QuantumToolkit.pulse(Unknown Source)
	at javafx.graphics@23/com.sun.javafx.tk.quantum.QuantumToolkit.pulseFromQueue(Unknown Source)
	at javafx.graphics@23/com.sun.javafx.tk.quantum.QuantumToolkit.lambda$runToolkit$11(Unknown Source)
	at javafx.graphics@23/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(Unknown Source)
	at javafx.graphics@23/com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
	at javafx.graphics@23/com.sun.glass.ui.gtk.GtkApplication.lambda$runLoop$10(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: java.net.URISyntaxException: Illegal character in query at index 58: https://go.gale.com/ps/i.do?p=HRCA&u=googlescholar&id=GALE|A653693412&v=2.1&it=r&sid=googleScholar&asid=68b1dec7
	at java.base/java.net.URI$Parser.fail(Unknown Source)
	at java.base/java.net.URI$Parser.checkChars(Unknown Source)
	at java.base/java.net.URI$Parser.parseHierarchical(Unknown Source)
	at java.base/java.net.URI$Parser.parse(Unknown Source)
	at java.base/java.net.URI.<init>(Unknown Source)
	... 53 more
```

</details>

### PR DESCRIPTION
**Pull request by hitalo-siriano**
Closes #11876.

- Introduced URLUtil.createUri() and URLUtil.create() to handle URL sanitization.
- Replaced direct calls to URI.create() and URI.create().toURL() with the new utility methods.
- URLs containing the pipe character ('|') are now properly encoded as '%7C' to prevent parsing errors.
- Added test cases to URLUtilTest to verify correct sanitization and URL creation.

### CODE REVIEW

**Comment by hitalo-siriano** [comment]
Am I on the right track to resolve this Issue?

File: src/main/java/org/jabref/gui/fieldeditors/URLUtil.java [hunk 1]
```diff
@@ -138,4 +143,17 @@ public static Optional<String> getSuffix(final String link, ExternalApplications
...
         }
     }
+
+    public static URL create(String url) throws MalformedURLException {
+        return URLUtil.createUri(url).toURL();
+    }
+
+    public static URI createUri(String url) {
+        try {
+            String urlFormat = url.replace("|", "%7C");
```
Comment (HoussemNasri) [inline src/main/java/org/jabref/gui/fieldeditors/URLUtil.java:153]: Please add a comment explaining why we are doing this and the relevant references.

File: src/main/java/org/jabref/gui/fieldeditors/URLUtil.java [hunk 2]
```diff
@@ -138,4 +143,17 @@ public static Optional<String> getSuffix(final String link, ExternalApplications
...
+            return new URI(urlFormat);
+        } catch (URISyntaxException e) {
```
Comment (koppor) [inline src/main/java/org/jabref/gui/fieldeditors/URLUtil.java:155]: Add `@throws` JavaDoc, too

**Review by koppor — COMMENTED** [review]
Add an Architecture Test disallowing the use of URI.create

**Comment by koppor** [comment]
@hitalo-siriano Are you still on it? The steps to finish it seem to be small.

**Comment by hitalo-siriano** [comment]
@koppor Yes, I'm still working on the changes

File: src/main/java/org/jabref/gui/fieldeditors/URLUtil.java [hunk 3]
```diff
@@ -143,13 +143,34 @@ public static Optional<String> getSuffix(final String link, ExternalApplications
             return Optional.ofNullable(suffix);
         }
     }
+    /**
+     * Creates a {@link URL} object from the given string URL.
+     *
+     * @param url the URL string to be converted into a {@link URL}.
+     * @return the {@link URL} object created from the string URL.
+     * @throws MalformedURLException if the URL is malformed and cannot be converted to a {@link URL}.
+     */
```
Comment (koppor) [inline src/main/java/org/jabref/gui/fieldeditors/URLUtil.java:153]: no empty line between JavaDoc and method

add an empty line above JavaDoc

**Review by koppor — CHANGES_REQUESTED** [review]

File: src/main/java/org/jabref/gui/fieldeditors/URLUtil.java [hunk 4]
```diff
@@ -62,7 +63,8 @@ public static String cleanGoogleSearchURL(String url) {
                 }
             }
             return url;
-        } catch (MalformedURLException e) {
+        } catch (
+                MalformedURLException e) {
```
Comment (koppor) [inline src/main/java/org/jabref/gui/fieldeditors/URLUtil.java:67]: Remove empty line (we know that the IntelliJ autoformatter creates this - but no one invested significant time to fix this)

**Comment by hitalo-siriano** [comment]
@koppor I made the requested changes, is there anything else that needs to be done?

**Comment by koppor** [comment]
@hitalo-siriano please see the automated output that your code misses matching our style rules. Fix that first. That reduces load on our side.

**Review by koppor — APPROVED** [review]

**Pull request merged**